### DISCUSSION TITLE
69. 정규화와 역정규화를 선택하는 기준을 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#71] [General]
정규화와 역정규화를 선택하는 기준을 설명해 주세요.

**Comment by eun-byeol** [comment]
정규화는 `데이터 무결성`과 `이상 현상을 방지(anomalies)`하기 위해 사용합니다. 데이터 중복을 최소화하여 구조 변경에 유연하고 쓰기 작업이 쉽습니다. 단, 테이블간 `조인(join)`으로 인해 복잡한 쿼리에서는 `읽기 성능이 저하`될 수 있습니다.

역정규화는 `읽기 성능을 개선`하고 `쿼리 복잡성을 줄이기` 위해 사용합니다. 정규화된 DB를 다시 중복을 포함하는 형태로 변경하여 쿼리 속도를 개선할 수 있습니다. 단, `데이터 무결성 유지가 어려울` 수 있고 `저장공간 비효율`이 발생할 수 있습니다.

**Comment by eun-byeol** [comment]
> 정규화(normalization)

관계형 DB에서 **데이터 중복을 최소화**하고 **데이터 일관성을 유지**하기 위해 데이터 구조를 조직화하는 프로세스

NF(normal form) : 정규화를 위한 룰. 1NF, 2NF, 3NF, BCNF, 4NF, 5NF, 6NF가 있음. 앞 단계를 만족해야 다음 단계 진행 가능

> 역정규화(denormalization)

**성능을 향상**시키기 위해 정규화된 DB를 다시 **중복을 포함하는 형태로 변경**하는 프로세스
